src/mito2/src/compaction/twcs.rs

```diff
@@ -0,0 +1,831 @@
+            region_id,
+            schema: region_metadata,
+            sst_layer: access_layer,
+            outputs,
+            expired_ssts,
+            sst_write_buffer_size: ReadableSize::mb(4),
+            compaction_time_window: None,
+            request_sender,
+            sender: waiters,
+            file_purger,
+        };
+        Ok(Some(Box::new(task)))
+    }
+}
+
+/// Assigns files to windows with predefined window size (in seconds) by their max timestamps.
+fn assign_to_windows<'a>(
+    files: impl Iterator<Item = &'a FileHandle>,
+    time_window_size: i64,
+) -> BTreeMap<i64, Vec<FileHandle>> {
+    let mut windows: BTreeMap<i64, Vec<FileHandle>> = BTreeMap::new();
+    // Iterates all files and assign to time windows according to max timestamp
+    for file in files {
+        let (_, end) = file.time_range();
+        let time_window = end
+            .convert_to(TimeUnit::Second)
+            .unwrap()
+            .value()
+            .align_to_ceil_by_bucket(time_window_size)
+            .unwrap_or(i64::MIN);
+        windows.entry(time_window).or_default().push(file.clone());
+    }
+    windows
+}
+
+/// Finds the latest active writing window among all files.
+/// Returns `None` when there are no files or all files are corrupted.
+fn find_latest_window_in_seconds<'a>(
+    files: impl Iterator<Item = &'a FileHandle>,
+    time_window_size: i64,
+) -> Option<i64> {
+    let mut latest_timestamp = None;
+    for f in files {
+        let (_, end) = f.time_range();
+        if let Some(latest) = latest_timestamp && end > latest {
+            latest_timestamp = Some(end);
+        } else {
+            latest_timestamp = Some(end);
+        }
+    }
+    latest_timestamp
+        .and_then(|ts| ts.convert_to_ceil(TimeUnit::Second))
+        .and_then(|ts| ts.value().align_to_ceil_by_bucket(time_window_size))
+}
+
+pub(crate) struct TwcsCompactionTask {
+    pub region_id: RegionId,
+    pub schema: RegionMetadataRef,
+    pub sst_layer: AccessLayerRef,
+    pub outputs: Vec<CompactionOutput>,
+    pub expired_ssts: Vec<FileHandle>,
+    pub sst_write_buffer_size: ReadableSize,
+    pub compaction_time_window: Option<i64>,
+    pub file_purger: FilePurgerRef,
+    /// Request sender to notify the worker.
+    pub(crate) request_sender: mpsc::Sender<WorkerRequest>,
+    /// Sender that are used to notify waiters waiting for pending compaction tasks.
+    pub sender: Option<Sender<error::Result<Output>>>,
+}
+
+impl Debug for TwcsCompactionTask {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TwcsCompactionTask").finish()
+    }
+}
+
+impl Drop for TwcsCompactionTask {
+    fn drop(&mut self) {
+        self.mark_files_compacting(false)
+    }
+}
+
+impl TwcsCompactionTask {
+    fn mark_files_compacting(&self, compacting: bool) {
+        self.outputs
+            .iter()
+            .flat_map(|o| o.inputs.iter())
+            .for_each(|f| f.set_compacting(compacting))
```

Comment:
issue: I realize that if we always set compacting to false here. There will be a race condition:
1. TwcsCompactionTask is finished
2. TwcsCompactionTask sends notification to the worker
4. TwcsCompactionTask drops and marks files as not compacting
5. The worker picks files to compact. It is possible to pick already compacted files to compact as compacting is false now
6. Then we may compact the same file twice.

Reply:
Region compacting status will be tracked inside `CompactionScheduler` to avoid two compaction tasks of a same region run concurrently.